devito/finite_differences/differentiable.py

```diff
@@ -263,7 +263,16 @@ def __neg__(self):
         return Mul(sympy.S.NegativeOne, self)
 
     def __eq__(self, other):
-        ret = super().__eq__(other)
+        from devito.finite_differences import Derivative
+        # Prevent recustion if other is also of derivative type
+        if type(self) is type(other) is Derivative:
+            # `sympy.Derivative.__eq__(self, other)` also recurses
+            ret = self.expr == other.expr
+            ret &= self.dimensions == other.dimensions
+            ret &= self.derivative_count == other.derivative_count
+            ret &= self.deriv_order == other.deriv_order
```

Comment:
This may or may not be sufficient, but I needed something in place to make progress

Reply:
should you not override `__eq__` inside `Derivative` instead?

Reply:
Yes, that is probably the right place, I think I fixed here as it was the source of the recursion exception. I think you probably still need something here to prevent infinite recursion